### PR TITLE
Clean up scrolling delegate code for 2D

### LINKED ISSUE #122650: Clean up scrolling delegates in prep for 2D
Part of the 2D scrolling proposal: [flutter.dev/go/2D-Foundation](https://github.com/flutter/flutter/issues/flutter.dev/go/2D-Foundation)

sliver.dart has a lot going on.
In the proposal for TwoDimensionalChildDelegate, it is going to live alongside the other scrolling delegate classes and share common code,
For ease of maintenance and readability, I'd like to move scrolling delegate logic into a separate file.

### PR DESCRIPTION
**Pull request by Piinks**
Fixes https://github.com/flutter/flutter/issues/122650

This is just a straight move (+ a bunch of templates to de-dupe a bunch of docs). I plan to add the two dimensional scrolling delegate base classes here as well. The `sliver.dart` file was very (very) large, and this neatly puts all the delegate code (including future 2D delegate in one place.)

This is motivated by part of the 2D scrolling proposal: [flutter.dev/go/2D-Foundation](https://github.com/flutter/flutter/pull/flutter.dev/go/2D-Foundation)

I will likely be re-using some common logic in the TwoDimensionalChildDelegates.

Similar to
- https://github.com/flutter/flutter/pull/122357

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Piinks** [comment]
I'd like to make a change here, please hold off on review.

**Comment by Hixie** [comment]
test-exempt: code refactor with no semantic change

**Comment by chunhtai** [comment]
Is this pr ready for review?

**Comment by Piinks** [comment]
> Is this pr ready for review?

Yes, thanks! This is ready now. :)

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 1]
```diff
@@ -11,872 +11,13 @@ import 'package:flutter/rendering.dart';
...
 import 'framework.dart';
-import 'selection_container.dart';
-
-export 'package:flutter/rendering.dart' show
-  SliverGridDelegate,
-  SliverGridDelegateWithFixedCrossAxisCount,
-  SliverGridDelegateWithMaxCrossAxisExtent;
+import 'scroll_delegate.dart';
 
 // Examples can assume:
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/sliver.dart:16]: Are there still examples in this file that need all of these (I noticed most of these were duplicated over in scroll_delegate.dart, wondering if some should be removed from here..)

File: packages/flutter/lib/src/widgets/scroll_delegate.dart [hunk 2]
```diff
@@ -0,0 +1,862 @@
...
+
+  @override
+  int? get estimatedChildCount => children.length;
+
+  @override
+  bool shouldRebuild(covariant SliverChildListDelegate oldDelegate) {
+    return children != oldDelegate.children;
+  }
+}
+class _SelectionKeepAlive extends StatefulWidget {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_delegate.dart:735]: nit: add a blank line before this class?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/sliver.dart (hunk 1 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/sliver.dart:16]: Oh right I meant to check that. Thank you!

File: packages/flutter/lib/src/widgets/scroll_delegate.dart (hunk 2 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/scroll_delegate.dart:735]:
```suggestion

class _SelectionKeepAlive extends StatefulWidget {
```

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**